lib/sqlalchemy/ext/mypy/plugin.py

```diff
@@ -217,80 +222,63 @@ def _cls_decorator_hook(ctx: ClassDefContext) -> None:
-    is encountered."""
-
+def _declarative_mixin_hook(ctx: ClassDefContext) -> None:
     _add_globals(ctx)
+    util.set_is_base(ctx.cls.info)
+    decl_class.scan_declarative_assignments_and_apply_types(
+        ctx.cls, ctx.api, is_mixin_scan=True
+    )
 
-    cls = ClassDef(ctx.name, Block([]))
-    cls.fullname = ctx.api.qualified_name(ctx.name)
-
-    info = TypeInfo(SymbolTable(), cls, ctx.api.cur_mod_id)
-    cls.info = info
-    _make_declarative_meta(ctx.api, cls)
-
-    cls_arg = util._get_callexpr_kwarg(ctx.call, "cls", expr_types=(NameExpr,))
-    if cls_arg is not None and isinstance(cls_arg.node, TypeInfo):
-        decl_class._scan_declarative_assignments_and_apply_types(
-            cls_arg.node.defn, ctx.api, is_mixin_scan=True
-        )
-        info.bases = [Instance(cls_arg.node, [])]
-    else:
-        obj = ctx.api.named_type("__builtins__.object")
-
-        info.bases = [obj]
 
-    try:
-        calculate_mro(info)
-    except MroError:
-        util.fail(
-            ctx.api, "Not able to calculate MRO for declarative base", ctx.call
-        )
-        obj = ctx.api.named_type("__builtins__.object")
-        info.bases = [obj]
-        info.fallback_to_any = True
+def _metaclass_cls_hook(ctx: ClassDefContext) -> None:
+    util.set_is_base(ctx.cls.info)
 
-    ctx.api.add_symbol_table_node(ctx.name, SymbolTableNode(GDEF, info))
 
+def _base_cls_hook(ctx: ClassDefContext) -> None:
+    _add_globals(ctx)
+    decl_class.scan_declarative_assignments_and_apply_types(ctx.cls, ctx.api)
 
-def _make_declarative_meta(
-    api: SemanticAnalyzerPluginInterface, target_cls: ClassDef
-) -> None:
 
-    declarative_meta_name: NameExpr = NameExpr("__sa_DeclarativeMeta")
-    declarative_meta_name.kind = GDEF
-    declarative_meta_name.fullname = "sqlalchemy.orm.decl_api.DeclarativeMeta"
+def _queryable_getattr_hook(ctx: AttributeContext) -> Type:
+    # how do I....tell it it has no attribute of a certain name?
+    # can't find any Type that seems to match that
+    return ctx.default_attr_type
 
-    # installed by _add_globals
-    sym = api.lookup_qualified("__sa_DeclarativeMeta", target_cls)
 
-    assert sym is not None and isinstance(sym.node, nodes.TypeInfo)
+def _add_globals(ctx: Union[ClassDefContext, DynamicClassDefContext]) -> None:
+    """Add __sa_DeclarativeMeta and __sa_Mapped symbol to the global space
+    for all class defs
 
-    declarative_meta_typeinfo = sym.node
-    declarative_meta_name.node = declarative_meta_typeinfo
+    """
 
-    target_cls.metaclass = declarative_meta_name
+    util.add_global(ctx, "sqlalchemy.orm.attributes", "Mapped", "__sa_Mapped")
 
-    declarative_meta_instance = Instance(declarative_meta_typeinfo, [])
 
+def _set_declarative_metaclass(
+    api: SemanticAnalyzerPluginInterface, target_cls: ClassDef
+) -> None:
     info = target_cls.info
-    info.declared_metaclass = info.metaclass_type = declarative_meta_instance
+    sym = api.lookup_fully_qualified_or_none(
```

Comment:
**mike bayer** (zzzeek) wrote:

hm I wonder why i didnt do this?  maybe i was emulating the old plugin that was actually doing something different.    good call

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2962